bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryFunctor.scala

```diff
@@ -59,43 +59,58 @@ class LchQueryFunctor(queryType: LchQueryFunctor.QueryType,
-          for {
-            gg  <- env.getGroups.asScala
-            gpt <- gg.getAll.asScala
-            if gpt.containsTarget(target)
-          } yield gpt.getGuider.getKey
-        }.headOption
-        else None
+    def makeTargetNode(skyObj: SPSkyObject, env: TargetEnvironment): Option[Serializable] = {
+      def extractGuiderKey(t: SPTarget): Option[String] = {
+        for {
+          gg  <- env.getGroups.asScala
+          gpt <- gg.getAll.asScala
+          if gpt.containsTarget(t)
+        } yield gpt.getGuider.getKey
+      }.headOption
+
+      def targetType: Option[String] = skyObj match {
+        case o: SPSkyObject if env.getSlewPositionObjectFromAsterism eq o => Some("Base")
+        case c: SPCoordinates                                             => Some(SPCoordinates.Tag)
+        case t: SPTarget    if env.isUserPosition(t)                      => Some("User")
+        case t: SPTarget    if env.isGuidePosition(t)                     => extractGuiderKey(t)
+        case _                                                            => None
       }
 
-      if (target.isNonSidereal) {
-        Some(new NonSidereal() {
-          setName(target.getName)
-          setType(targetType.orNull)
-          for {
-            n <- target.getNonSiderealTarget
-            h <- n.horizonsDesignation
-          } {
-            setHorizonsObjectId(h.show)
-          }
-        })
-      } else if (target.isSidereal) {
-        target.getSkycalcCoordinates(None.asGeminiOpt).asScalaOpt.map { coords =>
-          new Sidereal() {
-            setName(target.getName)
+      skyObj match {
+        case t: SPTarget if t.isNonSidereal =>
+          Some(new NonSidereal() {
+            setName(t.getName)
             setType(targetType.orNull)
-            setHmsDms(new HmsDms() {
-              setRa(HHMMSS.valStr(coords.getRa.getMagnitude))
-              setDec(DDMMSS.valStr(coords.getDec.getMagnitude))
-            })
+            for {
+              n <- t.getNonSiderealTarget
+              h <- n.horizonsDesignation
+            } {
+              setHorizonsObjectId(h.show)
+            }
+          })
+
+        case t: SPTarget if t.isSidereal =>
+          t.getSkycalcCoordinates(None.asGeminiOpt).asScalaOpt.map { coords =>
+            new Sidereal() {
+              setName(t.getName)
+              setType(targetType.orNull)
+              setHmsDms(new HmsDms() {
+                setRa(HHMMSS.valStr(coords.getRa.getMagnitude))
+                setDec(DDMMSS.valStr(coords.getDec.getMagnitude))
+              })
+            }
           }
-        }
-      } else None
+
+        case c: SPCoordinates =>
+          // For now, we pass off sky coordinates as a nonsidereal target to
+          // avoid having to significantly modify LTTS.
+          Some(new NonSidereal() {
+            setName(c.getName)
+            setType(targetType.orNull)
+          })
+
+        case _ => None
```

Comment:
Should we throw an exception here perhaps?  There are no other kinds of `SPSkyObject` (?) but if we did introduce one, we'd want to handle it and an exception would make it immediately obvious what is going on.

Reply:
You're right: the only two subclasses of `SPSkyObject` are `SPCoordinates` and `SPTarget`.

I think an exception is a good idea. I'll go ahead and change that.

Reply:
Oh, wait... it just occurred to me: there is one target type that we're not handling here, namely `TooTarget`, so it would throw exceptions for that. Is that what we would want?

Reply:
I'm not sure what should happen in that case.  It depends on how this is used but I'm not familiar with LCH.  We could make it explicit though and add a case for ToO targets that returns `None` (to match the previous behavior) and then an exception for anything unexpected.

Reply:
I'm going to do exactly that: `None` for `TooTarget` and throw an exception for the case that should never happen.